diff --git a/lua/nvim-highlight-colors/init.lua b/lua/nvim-highlight-colors/init.lua
index 9181f11..85f2f79 100644
--- a/lua/nvim-highlight-colors/init.lua
+++ b/lua/nvim-highlight-colors/init.lua
@@ -216,6 +216,7 @@ function M.turn_on()
 
 	is_loaded = true
 end
+M.turnOn = M.turn_on
 
 ---Callback to manually hide the highlights
 function M.turn_off()
@@ -227,6 +228,7 @@ function M.turn_off()
 
 	is_loaded = false
 end
+M.turnOff = M.turn_off
 
 ---Callback to manually toggle the highlights
 function M.toggle()
@@ -290,10 +292,4 @@ vim.api.nvim_create_user_command("HighlightColors",
 	}
 )
 
-return {
-	turnOff = M.turn_off,
-	turnOn = M.turn_on,
-	setup = M.setup,
-	toggle = M.toggle,
-	format = M.format,
-}
+return M
